src/components/menu/menu-popover.tsx

```diff
@@ -1,14 +1,79 @@
-import { FC, MutableRefObject, useRef } from 'react'
+import { FC, MutableRefObject, useEffect, useRef } from 'react'
 import { useClickOutside } from '../../hooks/use-click-outside'
 import { useMenuContext } from './menu-context'
 import { ElMenuPopover } from './styles'
 
+// close menu when focus move out of menu
+export const menuListFocusOutHandler = (menuList: HTMLDivElement, closeMenu: VoidFunction) => (event) => {
+  if (!menuList.contains((event as FocusEvent).relatedTarget as Node)) {
+    closeMenu()
+  }
+}
+
+// focus to the first menuitems when "arrowDown" pressed while `Menu` open
+export const menuButtonKeyDownHandler = (menuItems: NodeListOf<HTMLElement>) => (event) => {
+  if (event.key === 'ArrowDown') {
+    if (menuItems.length) menuItems[0].focus()
+  }
+}
+
+// handle basic navigation key refer to https://www.w3.org/WAI/ARIA/apg/patterns/menubar
+export const menuItemKeyDownHandler =
+  (menuButton: HTMLButtonElement, menuItems: NodeListOf<HTMLElement>, index: number, closeMenu: VoidFunction) =>
+  (event) => {
+    switch (event.key) {
+      case 'ArrowDown': {
+        const nextItem = menuItems[(index + 1) % menuItems.length]
+        nextItem.focus()
+        break
+      }
+      case 'ArrowUp': {
+        event.preventDefault()
+        const prevItem = menuItems[(index - 1 + menuItems.length) % menuItems.length]
+        prevItem.focus()
+        break
+      }
+      case 'Escape':
+        closeMenu()
+        menuButton.focus()
+        break
+      case 'Enter':
+      case ' ': {
+        const menuItem = menuItems[index]
+        menuItem.click()
+        break
+      }
+    }
+  }
+
 export const MenuPopover: FC = ({ children }) => {
   const { isOpen, closeMenu, getPopoverProps } = useMenuContext()
   const popoverRef = useRef<HTMLDivElement>(null)
 
   useClickOutside(popoverRef, closeMenu)
 
+  useEffect(() => {
```

Comment:
**issue:** It's really important we make the right unit of code responsible for the right behaviours, and I don't think it's quite right at the moment.

If we breakdown the responsibilities of each component within our Menu molecule, I think it would look something like this:

- `Menu.Item`: responsible for the presentational concerns of an item in the menu;
- `Menu.Group`: responsible for the presentational concerns of labelling a group of items in the menu;
- `Menu.Separator` (not implemented yet): responsible for the presentational concerns of a divider/separator in the menu;
- `Menu.List`: responsible for laying out menu items/groups in a vertical orientation;
- `Menu.Popper`: responsible for disclosing/hiding the menu list;
- `Menu.Trigger`: responsible for opening the menu when clicked;
- `useMenu`: responsible for providing the "glue" (state like `isOpen` and prop getters like `getTriggerProps`) that allow the menu atoms to function as a cohesive whole;
- `Menu`: responsible for integrating `useMenu` with `Menu.Context`.

Given this breakdown, is there a unit that stands out to you as the most appropriate place to implement extra behaviour (via event callbacks like `onKeyDown` and `onFocusOut`) we need for keyboard navigation?

Reply:
> issue: It's really important we make the right unit of code responsible for the right behaviours, and I don't think it's quite right at the moment.

thanks for describing this 👍 , I thought that it would be nice to centralize the keyboard handler logic but now I think It would be better to place it own related component

> Given this breakdown, is there a unit that stands out to you as the most appropriate place to implement the extra behaviour (via event callbacks like onKeyDown and onFocusOut) we need for keyboard navigation?

yes, I think that would be `Menu.Item` and `Menu.List` respectively

Reply:
> I think that would be getTriggerProps (for button keyDown), Menu.Item and Menu.List respectively, but I would like to try if Menu.List can handle both

Menu.List could be a reasonable choice and I won't block that approach if you feel strongly about it, but I'd suggest `useMenu` as an even better option.

My reasoning is:

- `Menu.List` is currently a presentational component. If we add keyboard handling to it, we no longer have a purely presentational component. This isn't necessarily bad, but..
- `useMenu` is already responsible for wiring up specific behaviour to specific atoms via its prop getters (like `getTriggerProps` providing an `onClick` to the menu trigger). It makes sense to do the same for the menu list (e.g. providing a `getMenuListProps` prop getter that wires up the necessary `onKeyDown`, `onFocusOut` callbacks). To put it another way, this fits in neatly with its existing responsibility as opposed to adding a new responsibility to `Menu.List`).
- By keeping behaviour inside `useMenu`, it's easy for us to swap out the presentational components with new ones in future without having to reimplement all the keyboard navigation/focus management behaviour.

Reply:
Ahh yes, I have overlooked the `getPopoverProps` 😞. I totally agree.